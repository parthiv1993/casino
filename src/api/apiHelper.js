const defaultHeaders = {
  Accept: 'application/json',
  'Content-Type': 'application/json'
};

export function API({ url, method, body, headers = {} }) {
  return {
    url,
    method,
    body: JSON.stringify(body),
    headers: { ...defaultHeaders, ...headers }
  };
}

export async function doAPICall({ url, method, body, headers }) {
  try {
    const result = await fetch(url, {
      method,
      body,
      headers
    });
    if (result.ok) {
      const jsonResult = await result.json();

      return jsonResult;
    } else {
      const jsonError = await result.json();
      throw new Error(jsonError.error);
    }
  } catch (error) {
    throw new Error(error);
  }
}
