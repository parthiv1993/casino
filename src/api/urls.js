export const baseURL = 'http://localhost:3001';

export const loginURL = `${baseURL}/login`;
export const logoutURL = `${baseURL}/logout`;
export const fetchGamesURL = `${baseURL}/games`;
export const fetchCategoriesURL = `${baseURL}/categories`;
