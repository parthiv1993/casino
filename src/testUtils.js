import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import light from './contexts/theme/light';

export const renderComponentWithLightTheme = (children) => {
  return render(<ThemeProvider theme={light}>{children}</ThemeProvider>);
};

export const lowerCaseAllKeys = (props = {}) => {
  const lowercaseObject = {};
  Object.keys(props).map((key) => (lowercaseObject[key.toLowerCase()] = props[key]));

  return lowercaseObject;
};

export const MockIcon = (props) => <div component="Icon" {...lowerCaseAllKeys(props)} />;
