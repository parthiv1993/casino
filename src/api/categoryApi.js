import { API, doAPICall } from './apiHelper';
import { fetchCategoriesURL } from './urls';

const fetchCategories = async () => {
  try {
    return await doAPICall(
      new API({
        url: fetchCategoriesURL,
        method: 'get'
      })
    );
  } catch (e) {
    throw new Error('SOMETHING_WENT_WRONG');
  }
};

export default fetchCategories;
