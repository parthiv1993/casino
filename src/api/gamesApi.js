import { API, doAPICall } from './apiHelper';
import { fetchGamesURL } from './urls';

const fetchGames = async () => {
  try {
    return await doAPICall(
      new API({
        url: fetchGamesURL,
        method: 'get'
      })
    );
  } catch (e) {
    throw new Error('SOMETHING_WENT_WRONG');
  }
};

export default fetchGames;
