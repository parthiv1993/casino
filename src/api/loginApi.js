import { API, doAPICall } from './apiHelper';
import { loginURL, logoutURL } from './urls';

export const login = async ({ username, password }) => {
  try {
    return await doAPICall(
      new API({
        url: loginURL,
        body: {
          username,
          password
        },
        method: 'post'
      })
    );
  } catch (e) {
    throw new Error('LOGIN_INVALID_CREDENTIALS');
  }
};

export const logout = async ({ username }) => {
  try {
    return await doAPICall(
      new API({
        url: logoutURL,
        body: {
          username
        },
        method: 'post'
      })
    );
  } catch (e) {
    throw new Error('SOMETHING_WENT_WRONG');
  }
};
