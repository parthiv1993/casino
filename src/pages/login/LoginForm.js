import { useCallback, useContext, useState } from 'react';
import { Button, Input } from '../../components';
import { validateInput } from './LoginForm.helpers';
import { StyledForm, ErrorMessage, LoginFormWrapper } from './LoginForm.styles';
import { login } from '../../api/loginApi';
import getLabel from '../../helpers/labels';
import getErrorMessages from '../../helpers/errorHelpers';
import LoginContext from '../../contexts/login/LoginContext';

const LoginForm = () => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(null);
  const { setUser } = useContext(LoginContext);

  const callLoginApi = useCallback(
    async (body) => {
      try {
        const response = await login(body);

        setUser({ ...response.player, username: body.username });
      } catch (e) {
        setError(e.message);
      }
    },
    [setUser]
  );

  const onFormSubmit = useCallback(
    async (event) => {
      event.preventDefault();
      const validationError = validateInput({ username, password });
      if (validationError) {
        setError(validationError);
      } else {
        await callLoginApi({ username, password });
      }
    },
    [callLoginApi, password, username]
  );

  return (
    <LoginFormWrapper>
      <StyledForm onSubmit={onFormSubmit}>
        <Input
          data-testid="username-input"
          suffixIcon="fa-solid fa-user"
          placeholder={getLabel('LOGIN_USERNAME_PLACEHOLDER')}
          type="text"
          onChange={setUsername}
          required
          name="username"
          autoFocus
        />
        <Input
          data-testid="password-input"
          suffixIcon="fa-solid fa-lock"
          placeholder={getLabel('LOGIN_PASSWORD_PLACEHOLDER')}
          type="password"
          name="password"
          onChange={setPassword}
          required
        />
        {error && (
          <ErrorMessage data-testid="error-message">{getErrorMessages(error)}</ErrorMessage>
        )}
        <Button type="submit" data-testid="submit-button">
          {getLabel('LOGIN_SUBMIT_BUTTON')}
        </Button>
      </StyledForm>
    </LoginFormWrapper>
  );
};

export default LoginForm;
