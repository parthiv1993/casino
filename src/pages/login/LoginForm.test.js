import LoginContext from '../../contexts/login/LoginContext';
import { renderComponentWithLightTheme } from '../../testUtils';
import { login } from '../../api/loginApi';
import LoginForm from './LoginForm';
import { fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

const mockSetUser = jest.fn();
const mockContextValue = { setUser: mockSetUser };

jest.mock('../../api/loginApi', () => ({
  ...jest.requireActual('../../api/loginApi'),
  login: jest.fn()
}));

const mockUser = {
  name: 'Parthiv',
  event: 'This is event',
  avatar: 'avatar/link'
};

describe('LoginForm', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    login.mockResolvedValue({ player: mockUser });
  });

  it('should match snapshot', () => {
    const { container } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should call login and set user in context when login is successfull', async () => {
    const { getByTestId, queryByTestId } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    fireEvent.change(getByTestId('username-input'), { target: { value: 'username1' } });
    fireEvent.change(getByTestId('password-input'), { target: { value: 'password1' } });
    await act(async () => {
      await fireEvent.click(getByTestId('submit-button'));
    });
    expect(queryByTestId('error-message')).toBeNull();
    expect(login).toHaveBeenCalledTimes(1);
    expect(login).toHaveBeenCalledWith({ username: 'username1', password: 'password1' });
    expect(mockSetUser).toHaveBeenCalledTimes(1);
    expect(mockSetUser).toHaveBeenCalledWith({ ...mockUser, username: 'username1' });
  });

  it('should not call login and show error message when username is empty', async () => {
    const { container, getByTestId } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    fireEvent.change(getByTestId('username-input'), { target: { value: '' } });
    fireEvent.change(getByTestId('password-input'), { target: { value: 'password1' } });
    await act(async () => {
      await fireEvent.click(getByTestId('submit-button'));
    });
    expect(getByTestId('error-message')).toBeDefined();
    expect(getByTestId('error-message')).toHaveTextContent('Please fill in username and passeord');
    expect(container).toMatchSnapshot();
    expect(login).toHaveBeenCalledTimes(0);
    expect(mockSetUser).toHaveBeenCalledTimes(0);
  });

  it('should not call login and show error message when username is invalid', async () => {
    const { getByTestId } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    fireEvent.change(getByTestId('username-input'), { target: { value: '$$$' } });
    fireEvent.change(getByTestId('password-input'), { target: { value: 'password1' } });
    await act(async () => {
      await fireEvent.click(getByTestId('submit-button'));
    });
    expect(getByTestId('error-message')).toBeDefined();
    expect(getByTestId('error-message')).toHaveTextContent(
      'Please fill in correct format username and password'
    );

    expect(login).toHaveBeenCalledTimes(0);
    expect(mockSetUser).toHaveBeenCalledTimes(0);
  });

  it('should not call login and show error message when password is empty', async () => {
    const { getByTestId } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    fireEvent.change(getByTestId('username-input'), { target: { value: 'username1' } });
    fireEvent.change(getByTestId('password-input'), { target: { value: '' } });
    await act(async () => {
      await fireEvent.click(getByTestId('submit-button'));
    });
    expect(getByTestId('error-message')).toBeDefined();
    expect(getByTestId('error-message')).toHaveTextContent('Please fill in username and passeord');
    expect(login).toHaveBeenCalledTimes(0);
    expect(mockSetUser).toHaveBeenCalledTimes(0);
  });

  it('should not call login and show error message when username is invalid', async () => {
    const { getByTestId } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    fireEvent.change(getByTestId('username-input'), { target: { value: 'username' } });
    fireEvent.change(getByTestId('password-input'), { target: { value: 'password1@@@' } });
    await act(async () => {
      await fireEvent.click(getByTestId('submit-button'));
    });
    expect(getByTestId('error-message')).toBeDefined();
    expect(getByTestId('error-message')).toHaveTextContent(
      'Please fill in correct format username and password'
    );

    expect(login).toHaveBeenCalledTimes(0);
    expect(mockSetUser).toHaveBeenCalledTimes(0);
  });

  it('should show error message in case api call fails', async () => {
    login.mockRejectedValue(new Error('LOGIN_INVALID_CREDENTIALS'));
    const { getByTestId } = renderComponentWithLightTheme(
      <LoginContext.Provider value={mockContextValue}>
        <LoginForm />
      </LoginContext.Provider>
    );
    fireEvent.change(getByTestId('username-input'), { target: { value: 'username' } });
    fireEvent.change(getByTestId('password-input'), { target: { value: 'password1' } });
    await act(async () => {
      await fireEvent.click(getByTestId('submit-button'));
    });
    expect(getByTestId('error-message')).toBeDefined();
    expect(getByTestId('error-message')).toHaveTextContent('Please enter valid credentials');

    expect(login).toHaveBeenCalledTimes(1);
    expect(mockSetUser).toHaveBeenCalledTimes(0);
  });
});
