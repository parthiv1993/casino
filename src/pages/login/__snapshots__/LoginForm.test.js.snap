// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match snapshot 1`] = `
.c3 {
  border: 0;
  outline: none;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background: transparent;
}

.c2 {
  width: 100%;
  border: 1px solid rgba(0,0,0,0.15);
  background-color: white;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5em;
  padding: 0.5em;
}

.c2.c2:focus-within {
  outline: 2px solid blue;
}

.c4 {
  background: white;
  border-width: 1px;
  border-color: rgba(0,0,0,0.15);
  color: #444;
  cursor: pointer;
  padding: 0.5em;
  border-radius: 4px;
}

.c4.c4:focus {
  outline: 2px solid blue;
}

.c0 {
  max-width: 400px;
  background-color: white;
  padding: 4rem 1rem;
  margin: auto;
  width: 100%;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <section
    class="c0"
  >
    <form
      class="c1"
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          data-testid="username-input"
          name="username"
          placeholder="Username"
          required=""
          type="text"
        />
        <i
          class="fa-solid fa-user"
        />
      </div>
      <div
        class="c2"
      >
        <input
          class="c3"
          data-testid="password-input"
          name="password"
          placeholder="Password"
          required=""
          type="password"
        />
        <i
          class="fa-solid fa-lock"
        />
      </div>
      <button
        class="c4"
        data-testid="submit-button"
        type="submit"
      >
        Login
      </button>
    </form>
  </section>
</div>
`;

exports[`LoginForm should not call login and show error message when username is empty 1`] = `
.c3 {
  border: 0;
  outline: none;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background: transparent;
}

.c2 {
  width: 100%;
  border: 1px solid rgba(0,0,0,0.15);
  background-color: white;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5em;
  padding: 0.5em;
}

.c2.c2:focus-within {
  outline: 2px solid blue;
}

.c5 {
  background: white;
  border-width: 1px;
  border-color: rgba(0,0,0,0.15);
  color: #444;
  cursor: pointer;
  padding: 0.5em;
  border-radius: 4px;
}

.c5.c5:focus {
  outline: 2px solid blue;
}

.c0 {
  max-width: 400px;
  background-color: white;
  padding: 4rem 1rem;
  margin: auto;
  width: 100%;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  color: red;
  margin: 0;
  text-align: center;
}

<div>
  <section
    class="c0"
  >
    <form
      class="c1"
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          data-testid="username-input"
          name="username"
          placeholder="Username"
          required=""
          type="text"
        />
        <i
          class="fa-solid fa-user"
        />
      </div>
      <div
        class="c2"
      >
        <input
          class="c3"
          data-testid="password-input"
          name="password"
          placeholder="Password"
          required=""
          type="password"
        />
        <i
          class="fa-solid fa-lock"
        />
      </div>
      <p
        class="c4"
        data-testid="error-message"
      >
        Please fill in username and passeord
      </p>
      <button
        class="c5"
        data-testid="submit-button"
        type="submit"
      >
        Login
      </button>
    </form>
  </section>
</div>
`;
